src/metabase/driver/mongo/query_processor.clj

```diff
@@ -39,26 +40,25 @@
 
 ;; # NATIVE QUERY PROCESSOR
 
-(defn- eval-raw-command
-  "Evaluate raw MongoDB javascript code. This must be ran insided the body of a `with-mongo-connection`.
-
-     (with-mongo-connection [_ \"mongodb://localhost/test\"]
-       (eval-raw-command \"db.zips.findOne()\"))
-     -> {\"_id\" \"01001\", \"city\" \"AGAWAM\", ...}"
-  [^String command]
-  (assert *mongo-connection* "eval-raw-command must be ran inside the body of with-mongo-connection.")
-  (let [^CommandResult result (.doEval ^DB *mongo-connection* command nil)]
-      (when-not (.ok result)
-        (throw (.getException result)))
-      (let [{result "retval"} (PersistentArrayMap/create (.toMap result))]
-        result)))
-
 (defn process-and-run-native
   "Process and run a native MongoDB query."
-  [_ query]
-  (let [results (eval-raw-command (:query (:native query)))]
-    (if (sequential? results) results
-        [results])))
+  [{{query :query} :native, database :database, :as outer-query}]
+  {:pre [query]}
+  (let [query   (u/prog1 (if (string? query)
+                           (json/parse-string query keyword)
+                           query)
+                  (println "QUERY ->" <>))
+        results  (mc/aggregate *mongo-connection* "users" query
+                               :allow-disk-use true)]
+    ;; As with Druid, we want to use `annotate` on the results of native queries.
+    ;; Because the Generic SQL driver was written in ancient times, it has its own internal implementation of `annotate`
+    ;; (which preserves the order of columns in the results) and as a result the `annotate` middleware isn't applied for `:native` queries.
+    ;; Thus we need to manually run it ourself here.
+    ;; TODO - it would be nice if we could let the middleware take care of `annotate` for us. This will likely require a new `IDriver` method
+    ;; returns whether or not `annotate` should be done to native query results.
+    (annotate/annotate outer-query (if (sequential? results)
```

Comment:
Given that the output structures between `process-native` and `process-structured` are clearly quite different, it seems like this is something that we could handle pretty easily.  In fact, since one is a map and the other is a collection then couldn't we just do something like this in the QP middleware function ...

```
(if-not (map? query-result)
 (annotate/annotate query-result)
 query-result)
```